[RFC] Lightrec timers

pcercuei opened the pull request:
Hi,

This RFC modifies PCSX-ReArmed to be able to utilize my (very) experimental "in-line" timers feature in Lightrec.

Basically, reading timers (and GPU status, but that's another story) is the main cause of I/O. Since those are handled in Lightrec by exiting the JIT and handling them in C in the glue, they are particularly costly.

My "in-line" timers optimization in Lightrec allows the glue to specify the start cycle and rate of the PSX timers. The generated code will then compute the value of a timer in-line, based on these values and the current cycle counter, in an optimized fashion.

Of course, this means that the cycle counter must not be reset every time; so I updated the glue to avoid that.

I only tested with a few games, and did not notice any bugs.

If you think it's OK, I will first merge my "timer" branch into Lightrec master, then update the PR to update Lightrec to the latest master branch.

notaz commented on libpcsxcore/lightrec/plugin.c:
```diff
@@ -171,6 +171,30 @@ static void lightrec_tansition_from_pcsx(struct lightrec_state *state)
+		reset_timer_data(state, (addr & 0x30) >> 4);
+		break;
+	}
+}
```
I think this can only work for timer 2 really. 0 and 1 have modes synced to the GPU where the timer is is reset on each scanline. Those are pretty much unemulated right now and will be needed for lightgun game support.

And does this work in general? Lightrec still counts 2 cycles/instruction, so the timer will be based on that? Does Parasite Eve II work with this?

Also what about about wrapping? Since psxRegs.cycle runs slower that lightrec's cycle counter, as soon as soon as it wraps won't there be a large timer jump observable by the game? For example with default settings the ratio is something like 1.14, so psxRegs.cycle counter value of ~0 ~= 613M lightrec's counter, then it wraps and cycles_pcsx_to_lightrec() does as well causing sudden jump to close to 0.

pcercuei replied:
Yes, wrapping is probably an issue. The reason I can't reset the counter every time is that the timers' `cycleStart` value is below psxRegs.cycle. But I can for instance reset Lightrec's cycle counter to e.g. 0x8000'0000 every time, and compute the "cycleStart" value that's passed to Lightrec accordingly. What matters are the offsets and not really the real values.

I did not know that timers 0 / 1 should reset on each scanline, but as you said, currently in PCSX it doesn't. If timers 0 / 1 need to reset on each scanline, I think it's still possible to handle it without jumping to C, using a modulo.

Right now the behaviour (except wrapping etc.) of Lightrec should be the exact same as calling the `psxRcntRcount` C function... in theory. When I was working on this optimization I still had a version of pcsx_rearmed without the CPU ratio implemented for Lightrec, and it matched 100%. Now with the cycles ratio implemented I'm struggling a bit (hence the RFC). I just tested Parasite Eve 2, and it does not work with this change. I guess because Lightrec still counts 2 cycles / instruction, yes.

One solution would probably be to use a 22.10 cycle counter in Lightrec, with a configurable cycles/op ratio in 22.10 format as well. Then the CPU ratio wouldn't have to be handled in the glue in the first place.

Btw, one problem about the CPU ratio approach is the loss of accuracy. Converting one value from PCSX to Lightrec scale then back to PCSX won't give you the exact same number.

pcercuei commented:
Updated, and rebased on top of PR #750.

This is still WIP, but now it works with Parasite Eve 2.

What puzzles me is the "Hack" commit, if I remove it, games don't get the button presses. Maybe there's one spot where the timers get updated and I don't sync with Lightrec's internal timers, and the hack acts as a magic work around, but I don't see it.

notaz commented:
I don't know how your timers work, especially considering lightrec_tansition_from_pcsx() resets lightrec's cycle counter, but it looks like lightrec's and the emu core's understanding of what the counter value should be diverges. Note that, for example, psxRegs.cycle may jump ahead in memhandlers for things like trying to compensate DMA stalling the CPU, causing the timers to jump ahead too, which I don't see how it could be reflected by lightrec's counters.

> I did not know that timers 0 / 1 should reset on each scanline

It's optional, there are bits for that in control register. Also timers 0 and 1 can use the pixel clock as source, which depends on the resolution and is not an integer multiple of CPU clock so the current rcnts[].rate can't really accommodate it. But maybe the fractional stuff you just added will be able to handle it.

> Btw, one problem about the CPU ratio approach is the loss of accuracy. Converting one value from PCSX to Lightrec scale then back to PCSX won't give you the exact same number.

I wouldn't worry about that since all the timing in this emulator is so incredibly inaccurate. There are 10+ cycle penalties on some IO, icache miss penalties end even RAM access penalties, all unemulated and generally wrong.

pcercuei commented:
Ok, it should work perfectly now.

I know that the psxRegs.cycle can jump ahead in memhandlers and such - but you repeating it pointed me to the right direction. I was having problems because of Lightrec's cycle counter being reset to a fixed value every time, which is a new setup to me as my local PCSX build doesn't do that. In my local build when psxRegs.cycle advances in memhandlers, the `(psxRegs.cycle - rcnts[timer].cycleStart)` difference also changes so Lightrec still computes correct values. With libretro's fork, I need to re-compute Lightrec's internal `cycleStart` value relative to the reset value of the cycle counter, every time the cycles counter is reset - aka. in `lightrec_tansition_from_pcsx` and before calling `lightrec_execute`.

pcercuei commented:
I did some benchmarks on GameCube today. The benefit seems to be of about 3 fps gained at the ~120 fps mark (117 avg. -> 120 avg.). Not really sure it's worth it.

notaz commented:
From a maintainer's POV multiple versions of the same code for each platform is a big PITA. Although if you limit it just for counter 2 it might be ok as that's least likely to need changes in future. Also, does it handle the reset-on-each-line mode needed for the lightguns?

pcercuei commented:
It does not handle the reset-on-each-line, it's barely faster and adds complexity. So I'll just give up on this one.
